addon/-private/properties/collection/main.js

```diff
@@ -0,0 +1,137 @@
+
+export class Collection {
+  constructor(definition, parent, key) {
+    this.definition = definition;
+    this.parent = parent;
+    this.key = key;
+
+    this._countPage = create({
+      count: count(definition.scope, { testContainer: definition.testContainer })
+    }, { parent });
+
+    this._items = [];
+  }
+
+  objectAt(index) {
+    if (this._items[index] === undefined) {
+      let { definition, parent, key } = this;
+      let scope = buildSelector({}, definition.scope, { at: index });
+
+      let finalizedDefinition = assign({}, definition);
+
+      finalizedDefinition.scope = scope;
+
+      let tree = create(finalizedDefinition, { parent });
+
+      // Change the key of the root node
+      Ceibo.meta(tree).key = `${key}(${index})`;
+
+      this._items[index] = tree;
+    }
+
+    return this._items[index];
+  }
+
+  get length() {
+    return this._countPage.count;
+  }
+
+  toArray() {
+    let { length } = this;
+
+    let array = Ember.A();
+
+    for (let i = 0; i < length; i++) {
+      array.push(this.objectAt(i));
+    }
+
+    return array;
+  }
+
+  map(...args) {
+    return this.toArray().map(...args);
+  }
+
+  mapBy(...args) {
+    return this.toArray().mapBy(...args);
+  }
+
+  forEach(...args) {
+    return this.toArray().forEach(...args);
+  }
+
+  findOne(query) {
+    let result = this.findAll(query);
+
+    if (result.length > 1) {
+      throwBetterError(
+        this.parent,
+        this.key,
+        `Expected at most one result from findOne query, but found ${result.length}`
+      );
+    }
+
+    return result[0];
+  }
+
+  findAll(query) {
+    let predicate;
+
+    if (typeof query === 'object') {
+      predicate = (item) => {
+        let isMatch = true;
+
+        for (let key in query) {
+          isMatch = isMatch && item[key] === query[key];
+        }
+
+        return isMatch;
+      }
+    } else if (typeof query === 'function') {
+      predicate = query;
+    } else {
+      throwBetterError(
+        this.parent,
+        this.key,
+        `Expected query for findAll to be either an object or function, received: ${query}`
+      );
+    }
+
+    return this.toArray().filter(predicate);
+  }
+}
+
+if (typeof (Symbol) !== 'undefined' && Symbol.iterator) {
+  Collection.prototype[Symbol.iterator] = function() {
+    let i = 0;
+    let items = this.toArray();
+    let next = () => ({ done: i >= items.length, value: items[i++] });
+
+    return { next };
+  }
+}
+
+export function collection(definition) {
+  let descriptor = {
+    isDescriptor: true,
+
+    setup(node, key) {
+      // Set the value on the descriptor so that it will be picked up and applied by Ceibo.
+      // This does mutate the descriptor, but because `setup` is always called before the
+      // value is assigned we are guaranteed to get a new, unique Collection instance each time.
+      descriptor.value = new Collection(this._definition, node, key);
+    },
+
+    _definition: definition
```

Comment:
Ah, so this is actually completely for the benefit of `ember-classy-page-object` I just realized, so I would definitely be open to changing it or removing it.

The key issue here is that because collections are a function and return a descriptor, there is no way to access the definition externally. So, you can't extend a collection's definition later on. Basically, there's no way to do:

```js

let pageOne = {
  items: collection({
    scope: 'li'
  })
}

let pageTwo = {
  items: collection({
    ...pageOne.items,

    click: clickable()
  })
}
```

This would make that definition accessible later on. There are workarounds for this that I can use in `ember-classy-page-object`, though they aren't super pretty. Either way, I'm open.

Reply:
Yah, I think I'd prefer to remove for now and tackle extending collections as a separate issue, just to reduce the problems to solve & reason about in this PR. It's an interesting issue that probably deserves its own discussion.